fix #602 Changes to preserve quotes on arguments commands so rundeck will interpret it as unique argument

carlosrfranco opened the pull request:
fixes #602 

**Is this a bugfix, or an enhancement? Please describe.**
If a command has a quoted strings like `command -m 'arg1 arg2'`, this arguments was not preserved by rundeck and the argument `'arg1 arg2'`  was interpreted as two separated args.

**Describe the solution you've implemented**
Quotes will be preserved and is interpreted as unique argument

gschueler commented on rundeckapp/grails-app/views/execution/_wfitemEdit.gsp:
```diff
@@ -475,6 +475,21 @@
+                                            checked="${item?.preserveQuotesOnArguments}"
+                                            id="preserveQuotesOnArgumentsCheck" value="true"
+                                />
+                                <label for="preserveQuotesOnArguments">
```
incorrect value prevents label from working
```suggestion
                                <label for="preserveQuotesOnArgumentsCheck">
```

carlosrfranco replied:
Fixed

carlosrfranco commented:
> I think we need some additional clarity on the behavior:
> 
>     * preserving quotes should be the default behavior when creating new commands.  however it should should be off for existing commands for compatibility, and the checkbox to toggle it should also explain a bit more about why you might want to disable it for compatibility.
> 
I added a tooltip with additional explaining for the behavior.

![Captura de ecrã de 2020-01-03 11-34-37](https://user-images.githubusercontent.com/7209908/71729273-0706bc80-2e1e-11ea-87d5-b6a196c4fc16.png)

> 
> Commands run from the "Commands" section work in the old (non-preserving) way. Perhaps another toggle needs to be added to Commands UI to change this as well

Added on last commit

ahonor commented:
Is there a feature flag that can toggle this feature off/on?

carlosrfranco commented:
> Is there a feature flag that can toggle this feature off/on?

In this current state does not have the feature flag. I understand that the idea is to keep the previous behavior for existing jobs so even without the feature flag, the jobs will not be affected. But of course it wouldn't be a problem to add the flag as well

gschueler commented:
@carlosrfranco  @ahonor yes let's add a feature toggle. if turned off, the old behavior is used and the checkbox for command steps and commands tab is not shown.  if turned on, the new behavior is default, but checkboxes are shown.  I dislike the checkboxes, but not sure of a better way to allow the new behavior to be default and still have compatibility for old job definitions

gschueler commented on rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/api/ApiRunAdhocRequest.groovy:
```diff
@@ -37,10 +37,12 @@ class ApiRunAdhocRequest implements Validateable{
     String fileExtension
     Boolean nodeKeepgoing
     Integer nodeThreadcount
+    Boolean preserveQuotesOnArgumentsCheck
```
suggest renaming this to `preserveQuotes`

gschueler replied:
also, as this now affects the API, the `apiRunCommand` in ScheduledExecutionController should support this parameter only in newer API version.

gschueler replied:
I think also, given newer API version, the run command API should use preserveQuotes=true by default if it is not specified, otherwise set it to false.

ahonor replied:
@carlosrfranco yes, it should be toggled off by default.

carlosrfranco commented:
@gschueler Will the feature flag be off by default?

carlosrfranco commented:
PR updated and moved to https://github.com/rundeck/rundeck/pull/6887
